sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1396 @@
+graph is undefined. If user code enforces a total order on the queue
+events, then the behavior is well-defined, and will match the observable
+total order.
+
+The returned value from the `info::queue::state` should be considered
+immediately stale in multi-threaded usage, as another thread could have
+preemptively changed the state of the queue.
+
+=== Exception Safety
+
+In addition to the destruction semantics provided by the SYCL
+{crs}[common reference semantics], when a modifiable `command_graph` is
+destroyed recording is ended on any queues that are recording to that
+graph, equivalent to `this->end_recording()`.
+
+As a result, users don't need to manually wrap queue recording code in a
+`try` / `catch` block to reset the state of recording queues on an exception
+back to the executing state. Instead, an uncaught exception destroying the
+modifiable graph will perform this action, useful in RAII pattern usage.
+
+=== Error Handling
+
+Errors are reported through exceptions, as usual in the SYCL API. For new APIs,
+submitting a graph for execution can generate unspecified asynchronous errors,
+while `command_graph::finalize()` may throw unspecified synchronous exceptions.
+Synchronous exception errors codes are defined for all of
+`command_graph::add()`, `command_graph::make_edge()`, `command_graph::update()`,
+`command_graph::begin_recording()`, and `command_graph::end_recording()`.
+
+Submitting an executable graph using `handler::ext_oneapi_graph()` to
+a queue with a different SYCL context than that of the executable graph will
+result in a synchronous exception.
+
+When a queue is in recording mode asynchronous exceptions will not be
+generated, as no device execution is occurring. Synchronous errors specified as
+being thrown in the default queue executing state, will still be thrown when a
+queue is in the recording state. A synchronous error with error code `invalid`
+will also be thrown if a queue in recording mode tries to record a command to a
+graph which already has nodes that target a device different from the device
+associated with the recording queue.
+
+The `command_graph::begin_recording` and `command_graph::end_recording`
+entry-points return a `bool` value informing the user whether a related queue
+state change occurred. False is returned rather than throwing an exception when
+no queue state is changed. This design is because the queues are already in
+the state the user desires, so if the function threw an exception in this case,
+the application would likely swallow it and then proceed.
+
+While a queue is in the recording state, methods performed on that queue which
+are not command submissions behave as normal except for waits. Waiting on a
+queue in the recording state is an error and will throw a synchronous
+exception. Other methods are ignored by the graph system as opposed to
+throwing in recording mode. As any query about the state of the queue may
```

Comment:
What does this mean about other methods being ignored?  I thought the general behavior for record mode is that a command either behaves as normal, or it fails with an exception.  When considering "behaves as normal", you need to consider only the observable effect.  For example, `queue::submit` does a different thing in record mode, but an application can't observe any difference between scheduling a command for future execution vs. recording a command for future execution.

I think it would be good to have a single section that specifies exactly which queue operations fail when the queue is in record mode.  This would contain information that is currently spread around in different places, such as the prohibition on waiting and the "Buffer Limitations for Record & Replay".  I think this section should list specific APIs that fail.  For those that only fail conditionally, it should describe exactly what that condition is.

Reply:
You're intent, then, is to say that member functions like `get_backend` are unaffected by record mode and operate as usual?  This is good, but the wording is confusing:

> Other methods are ignored by the graph system as opposed to throwing in recording mode.

I think it would be clearer if there was a separate section describing the behavior that changes when a queue is in record mode.  Something like:

> === Queue behavior in record mode
>
> When a queue is placed in record mode via a call to `command_graph::begin_recording`, some features of the queue are no longer available because the commands are not executed during this mode.  The general philosophy is to throw an exception at runtime when a feature is not available, so that there is an obvious indication of failure.  The following list describes the behavior that changes during record mode.  Features not listed below behave the same in record mode as they do in non-record mode.
> 
> [List of changes here]

Reply:
>   Other methods are ignored by the graph system as opposed to throwing in recording mode.

Rephrased this in https://github.com/reble/llvm/pull/133 to say more clearly that query methods on a queue work as expected